### ISSUE TITLE
[Question] How can I get the categorical distribution as actions instead of the samples ?

### ISSUE THREAD

**Issue by sabeaussan** [DLR-RM/stable-baselines3#1777]
### ❓ Question

My question is simply the following : I need to retrieve the categorical distribution outputted by the actor before sampling (right before the softmax layer) to act rather than a sample of that distribution but the action returned during learning is sample. I looked into the doc and it didn't seem to have an easy way to achieve that. Any advice regarding this ?

Thank !

### Checklist

- [X] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo
- [X] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/)
- [X] If code there is, it is [minimal and working](https://github.com/DLR-RM/stable-baselines3/issues/982#issuecomment-1197044014)
- [X] If code there is, it is formatted using the [markdown code blocks](https://help.github.com/en/articles/creating-and-highlighting-code-blocks) for both code and stack traces.

**Comment by qgallouedec** [comment]
A bit tricky, but here is a way

```python
from stable_baselines3 import DQN
from stable_baselines3.common.env_util import make_atari_env
from stable_baselines3.common.preprocessing import maybe_transpose
from stable_baselines3.common.utils import obs_as_tensor

# Create the environment and the model
env = make_atari_env("BreakoutNoFrameskip-v4", n_envs=2, seed=0)
model = DQN("MlpPolicy", env, verbose=1)

# Take any action
obs = env.reset()

# Preprocess the observation
obs = maybe_transpose(obs, model.observation_space)
obs_tensor = obs_as_tensor(obs, model.device)

# Compute the logits
logits = model.policy.q_net(obs_tensor)  # (n_envs, n_actions)
```

**Comment by araffin** [comment]
Duplicate of https://github.com/DLR-RM/stable-baselines3/issues/1257 and https://github.com/DLR-RM/stable-baselines3/issues/1347

DQN: https://github.com/DLR-RM/stable-baselines3/issues/568

**Comment by sabeaussan** [comment]
Thank you for the answer ! I'm struggling to comprehend how to implement this method in the training loop. Since everything is happening inside the "learn" method, I don't see how this can be integrated during training. I already considered adding an Action Wrapper that implement this but I think this is sub-optimal since it would trigger two forward passes (one before the wrapper and one to compute the logits from observation). I'm also considering installing sb3 from source and making the modifications there but a bit more time consumming. Any advice on this matter ?

**Comment by sabeaussan** [comment]
I mean that during training (i.e when I call the "learn" method on the model) the action send to the "step" function of my custom env should be the categorical distribution and not the sampled action. More context : 

I'm building a resource allocation environment where an agent is tasked to allocate limited resources based on demands and some constraints. The allocation of resource should be distributed according to the categorical distribution derived from the logits. In my case I have an action space Discrete(25), and after calling "learn" on PPO or DQN, it should step my env with a vector of dimension 25 containing the norm prob, rather than a single sample.
